File: bindings/python/tests/targeted.py
from cellrune import CellRuneError, TargetCalculationResult, Workbook


def main() -> None:
    with Workbook.create() as workbook:
        workbook.set_number("Sheet1", "A1", 2)
        workbook.set_formula("Sheet1", "B1", "=A1+1")
        workbook.set_formula("Sheet1", "C1", "=B1*2")
        workbook.set_formula("Sheet1", "Z1", "=UNKNOWN_FUNCTION(1)")
        targets = [{"sheet": "Sheet1", "start": "B1", "end": "D1"}]
        result: TargetCalculationResult = workbook.calculate_targets(targets)
        assert result["scope"] == "targets"
        assert result["semantic_revision"] == workbook.summary()["semantic_revision"]
        assert result["evaluated_count"] == 2
        assert result["parsed_formula_count"] == 2
        assert [item["result"] for item in result["cells"]] == [
            {"kind": "value", "value": {"kind": "number", "value": 3}},
            {"kind": "value", "value": {"kind": "number", "value": 6}},
            {"kind": "value", "value": {"kind": "blank"}},
        ]
        assert workbook.read_range("Sheet1", "C1", "C1")["cells"][0]["calculated"] is None
        assert not workbook.changes_since()["deltas"]
        try:
            workbook.to_bytes()
        except CellRuneError as error:
            assert error.code == "interop.calculation.required"
        else:
            raise AssertionError("partial result enabled saving")
        workbook.recalculate()
        cached = workbook.calculate_targets(targets)
        assert cached["reused_count"] == 2
        assert cached["evaluated_count"] == 0
        workbook.set_number("Sheet1", "A1", 5)
        edited = workbook.calculate_targets([{"sheet": "Sheet1", "start": "C1"}])
        assert edited["cells"][0]["result"] == {"kind": "value", "value": {"kind": "number", "value": 12}}
        assert edited["reused_count"] == 0
        for invalid in ([], [{"sheet": "Sheet1", "start": "A1", "end": "XFD1048576"}]):
            try:
                workbook.calculate_targets(invalid)
            except CellRuneError:
                pass
            else:
                raise AssertionError("invalid targets accepted")
        try:
            workbook.calculate_targets(targets, limits={"max_result_cells": True})
        except CellRuneError:
            pass
        else:
            raise AssertionError("boolean limit accepted")


if __name__ == "__main__":
    main()

File: crates/cellrune-interop/src/targeted.rs
//! Transport contract for immutable, bounded partial calculation.

use crate::{
    CalculationOptionsDto, CalculationOptionsReportDto, CalculationResultDto, CellReferenceDto,
    ProviderIdentityDto, WorkbookFingerprintDto,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// One cell or inclusive rectangle to calculate on a named sheet.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct CalculationTargetDto {
    /// Case-insensitive sheet name.
    pub sheet: String,
    /// Unqualified A1 address of the first cell.
    pub start: String,
    /// Last cell; omission requests only `start`.
    #[serde(default)]
    pub end: Option<String>,
}

/// Request-wide bounds, separate from the formula kernel limits.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(default, deny_unknown_fields)]
pub struct TargetCalculationLimitsDto {
    /// Maximum number of input rectangles. Default: 1024.
    pub max_targets: u32,
    /// Maximum distinct returned cells. Default: 10000.
    pub max_result_cells: u32,
    /// Maximum evaluator invocations, including dynamic retries. Default: 100000.
    pub max_evaluated_cells: u32,
}

impl Default for TargetCalculationLimitsDto {
    fn default() -> Self {
        Self {
            max_targets: 1024,
            max_result_cells: 10000,
            max_evaluated_cells: 100000,
        }
    }
}

/// Explicit targets and deterministic inputs for one partial calculation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct TargetCalculationRequestDto {
    /// Requested cells and rectangles; overlap is deduplicated.
    pub targets: Vec<CalculationTargetDto>,
    /// Deterministic numeric and volatile-function inputs.
    #[serde(default)]
    pub options: CalculationOptionsDto,
    /// Request-wide resource limits.
    #[serde(default)]
    pub limits: TargetCalculationLimitsDto,
}

/// Explicit marker that the result covers requested cells only.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum TargetCalculationScopeDto {
    /// Requested cells, with required precedents evaluated internally.
    Targets,
}

/// One requested cell value or issue.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct TargetCalculationCellDto {
    /// Stable identity and current address of the requested cell.
    pub cell: CellReferenceDto,
    /// Calculated value or stable unavailable issue.
    pub result: CalculationResultDto,
}

/// Complete bounded response to a partial calculation, never an installed workbook cache.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct TargetCalculationResultDto {
    /// Version of the serialized interop contract.
    pub schema_version: u32,
    /// Explicit partial-result scope.
    pub scope: TargetCalculationScopeDto,
    /// Revision of the immutable source workbook.
    pub semantic_revision: u64,
    /// Semantic identity of the source workbook.
    pub source_fingerprint: WorkbookFingerprintDto,
    /// Original package hash, when available.
    pub input_sha256: Option<String>,
    /// Calculator identity and version.
    pub calculator_provider: ProviderIdentityDto,
    /// Complete deterministic options and formula limits.
    pub calculation_options: CalculationOptionsReportDto,
    /// Request-wide resource limits.
    pub limits: TargetCalculationLimitsDto,
    /// Requested cells, deduplicated and sorted by sheet ID, row, then column.
    pub cells: Vec<TargetCalculationCellDto>,
    /// Evaluator invocations, including precedents and dynamic retries.
    pub evaluated_count: u64,
    /// Formula cells parsed for this request.
    pub parsed_formula_count: u64,
    /// Requested materialized results reused from a current complete calculation.
    pub reused_count: u64,
}
